aspnet/migration/http-modules.rst

```diff
@@ -1,8 +1,588 @@
-.. include:: /../common/stub-topic.txt
+Migrating HTTP Modules to Middleware
+====================================
 
-|stub-icon| Migrating HTTP Modules to Middleware
-======================================================
+By `Matt Perdeck`_
 
-.. include:: /../common/stub-notice.txt
+`Modules and HTTP Handlers <https://msdn.microsoft.com/en-us/library/bb398986.aspx>`_
+are no longer available in ASP.NET 5. 
+This article shows how to migrate your modules and handlers to the new ASP.NET 5 :ref:`middleware <fundamentals-middleware>`.
+
+.. contents:: In this article:
+  :local:
+  :depth: 1
+
+Handlers and modules revisited
+------------------------------
+Before proceeding to ASP.NET 5 middleware, let's first recap the use of handlers and modules in ASP.NET 4 and below: 
+
+.. image:: http-modules/_static/moduleshandlers.png
+
+Handlers are:
+	* Classes that implement `IHttpHandler <https://msdn.microsoft.com/en-us/library/system.web.ihttphandler(v=vs.100).aspx>`_;
+	* Used to handle only requests with a given file name or extension, such as *.report*;
+	* `Configured <https://msdn.microsoft.com/en-us/library/46c5ddfy(v=vs.100).aspx>`__ in your *web.config* file.
+
+Modules are:
+	* Classes that implement `IHttpModule <https://msdn.microsoft.com/en-us/library/system.web.ihttpmodule(v=vs.100).aspx>`_;
+	* Invoked in principle for every request;
+	* Able to stop further processing of a request;
+	* Able to add to the HTTP response, or create their own; 
+	* `Configured <https://msdn.microsoft.com/en-us/library/ms227673(v=vs.100).aspx>`__ in your web.config file.
+
+**The order in which modules process incoming requests is determined by:**
+
+	1.	The `application life cycle <https://msdn.microsoft.com/en-us/library/ms227673(v=vs.100).aspx>`_, 
+		which is a series events fired by ASP.NET
+		- 
+		`BeginRequest <https://msdn.microsoft.com/en-us/library/system.web.httpapplication.beginrequest(v=vs.100).aspx>`_,
+		`AuthenticateRequest <https://msdn.microsoft.com/en-us/library/system.web.httpapplication.authenticaterequest(v=vs.100).aspx>`_, etc.
+		Each module can create a handler for one or more events;
+
+	2. When handling the same event, the order in which they are configured in the web.config file.
+
+In addition to modules, you can add handlers for the life cycle events to your *Global.asax.cs* file. These handlers run after the handlers in the configured modules.
+
+From handlers and modules to middleware
+---------------------------------------
+Handlers and modules do the job, but are overly complicated:
+	* Seeing that modules can create responses themselves, there is no need for handlers;
+	* The order in which modules run can be determined solely by the order in which they are inserted into the request pipeline, making the life cycle events superfluous;
```

Comment:
This isn't quite true. The HTTP module pipeline allowed modules to execute in an *interleaved* fashion. You can't do that with a single middleware. In fact, it has often been very complicated to build certain types of middleware due to this limitation. Two examples of this are the ELM feature (Error Logging Middleware), which needs one middleware for logging, and one middleware for viewing results, and another example is Application Insights, which needs one middleware for timing, and one for error reporting. With HTTP modules you'd need only one.